src/smt/proof_post_processor_dsl.cpp

```diff
@@ -109,7 +60,34 @@ bool ProofPostprocessDsl::shouldUpdate(std::shared_ptr<ProofNode> pn,
                                        bool& continueUpdate)
 {
   ProofRule id = pn->getRule();
-  return id == ProofRule::TRUST || id == ProofRule::TRUST_THEORY_REWRITE;
+  continueUpdate = true;
+  // we should update if we
+  // - Have rule TRUST or TRUST_THEORY_REWRITE,
+  // - We have no premises
+  // - We are not already recursively expanding >= 3 steps of the above form.
+  // We check for the third criteria by tracking a d_traversing vector.
+  if ((id == ProofRule::TRUST || id == ProofRule::TRUST_THEORY_REWRITE)
+      && pn->getChildren().empty() && d_traversing.size() < 3)
+  {
+    Trace("pp-dsl-process") << "...push " << pn.get() << std::endl;
+    d_traversing.push_back(pn);
```

Comment:
Hopefully, the traversal is depth-first and not breadth-first.

Reply:
Yes, it is.